src/osd/PrimaryLogPG.cc

```diff
@@ -4458,8 +4531,37 @@ int PrimaryLogPG::trim_object(
     ctx->delta_stats.num_object_clones--;
     if (coi.is_cache_pinned())
       ctx->delta_stats.num_objects_pinned--;
-    if (coi.has_manifest())
+    if (coi.has_manifest()) {
+      set<uint64_t> refs;
+      object_info_t oi;
+      for (auto p : snapset.clones) {
+	hobject_t clone_oid = coid;
+	if (clone_oid.snap == p) {
+	  continue;
+	}
+	clone_oid.snap = p;
+	ObjectContextRef cobc = get_object_context(clone_oid, false, NULL);
+	if (!cobc) {
+	  close_op_ctx(ctx.release());
+	  dout(0) << __func__ << ": Can not find obc " << coid << dendl;;
+	  return -ENOENT;
+	}
+
+	// check if the references is still used
+	oi = cobc->obs.oi;
+	if (oi.has_manifest()) {
+	  refs = coi.manifest.build_intersection_set(oi.manifest.chunk_map);
```

Comment:
I apologize, I failed to notice that refs got passed in for each clone.  Unfortunately, that makes this version wrong -- you need the union of all refs in common with any clone.  I suggest you swap it back to passing in refs by reference.  I'd also like you to add a test that validates this behavior (at the moment, it'll drop references in common with all but the last clone).

There's another problem I didn't notice before.  Consider the sequence:
```
write foo [0, 1024)
flush foo ->
  head: [0, 512) aaa, [512, 1024)  bbb
  refcount(aaa)=1, refcount(bbb)=1
snapshot 10
write foo [0, 512) ->
  head:               [512, 1024)  bbb
  10  : [0, 512) aaa, [512, 1024) bbb
  refcount(aaa)=1, refcount(bbb)=1
flush foo ->
  head: [0, 512) ccc, [512, 1024)  bbb
  10  : [0, 512) aaa, [512, 1024) bbb
  refcount(aaa)=1, refcount(bbb)=1, refcount(ccc)=1
snapshot 20
write foo [0, 512) (same contents as the original write)
  head:               [512, 1024)  bbb
  20  : [0, 512) ccc, [512, 1024) bbb
  10  : [0, 512) aaa, [512, 1024) bbb
  refcount(aaa)=?, refcount(bbb)=1
flush foo
  head: [0, 512) aaa, [512, 1024)  bbb
  20  : [0, 512) ccc, [512, 1024) bbb
  10  : [0, 512) aaa, [512, 1024) bbb
  refcount(aaa)=?, refcount(bbb)=1, refcount(ccc)=1
```

What should the refcount be here?  I claim that the refcount should be 2.  We don't want to have to check *every* clone when we set or clear a clone -- we really only want to check adjacent clones.  However, this creates an additional complication when trimming 20:

```
initial:
  head: [0, 512) aaa, [512, 1024)  bbb
  20  : [0, 512) ccc, [512, 1024) bbb
  10  : [0, 512) aaa, [512, 1024) bbb
  refcount(aaa)=2, refcount(bbb)=1, refcount(ccc)=1
trim 20
  head: [0, 512) aaa, [512, 1024)  bbb
  10  : [0, 512) aaa, [512, 1024) bbb
  refcount(aaa)=?, refcount(bbb)=1, refcount(ccc)=0
```

By our rule above, refcount(aaa) must now be 1, so trimming clone 20 resulted in decrementing the refcount for a reference it doesn't even have.  It's worth at this point reminding ourselves of why we are permitting this complexity:
1. We don't want to burden make_writeable with the need to increment a refcount prior to committing the new obc (hot path for writes), so  it's important that we adopt a convention such that adding a clone between HEAD and the most recent clone that already exists never increments the refcount.
2. We don't want to have to scan all clones to determine whether a reference we are adding already exists -- thus it's sensible to only consider adjacent clones.

The above two rules don't require specifically this approach, but I think any design we adopt needs to satisfy them.

If you want, I can write up a patch to add to this PR to address this next week.

Note, flush will need to deal with this as well by checking adjacent clones -- if the clones on either side already have the reference, adding to the middle one would actually result in decrementing the refcount.

Reply:
I slipped up. Fixed. So, what is the key idea to avoid scanning all clones?

Reply:
The convention I outlined above solves it -- consecutive clones with the same ref in the same offset share a single ref.  It just means that here you need to check for the case I outlined: refs on both chunks on either side but not on the clone being removed.

Reply:
I'd expect you'll want to replace build_intersection_set with something like find_refs_to_drop(manifest_t *g, manifest *l) which takes the manifests for clones immediately greater than (g) and less than (l) (nullptr on g for head, l for the oldest clone etc), and outputs the set of refs to drop.